packages/wallet-sdk/src/core/communicator/Communicator.ts

```diff
@@ -96,5 +110,40 @@ export class Communicator {
         if (!this.popup) throw standardErrors.rpc.internal();
         return this.popup;
       });
-  }
+  };
+
+  /**
+   * Posts a RPC request to the popup window and waits for a response
+   */
+  postRPCRequest = (request: RPCRequestMessage): Promise<RPCResponseMessage> => {
+    this.pendingRequests.set(request.id, 'created');
+
+    return new Promise((resolve, reject) => {
+      // defer the execution the promise until the current call stack clears
+      Promise.resolve().then(() => {
+        // reject if the request was cancelled between creation and awaiting due to PopupUnload
+        if (!this.pendingRequests.has(request.id)) {
+          reject(standardErrors.provider.userRejectedRequest('Request cancelled before sending'));
+          return;
+        }
+
+        // otherwise, post the message and wait for a response
+        this.pendingRequests.set(request.id, 'awaiting');
+        this.postMessage(request)
+          .then(() =>
+            this.onMessage<RPCResponseMessage>(({ requestId }) => requestId === request.id)
```

Comment:
`onMessage` can be cancelled as well, if popup gets closed while waiting for response. 
![image](https://github.com/coinbase/coinbase-wallet-sdk/assets/966662/37f5b765-a3fa-4f1b-9111-c401b8d36aa1)